pymc3/distributions/distribution.py

```diff
@@ -134,40 +234,165 @@ def __new__(cls, name, *args, **kwargs):
                 "for a standalone distribution."
             )
 
-        rng = kwargs.pop("rng", None)
-
-        if rng is None:
-            rng = model.default_rng
-
         if not isinstance(name, string_types):
             raise TypeError(f"Name needs to be a string but got: {name}")
 
-        data = kwargs.pop("observed", None)
+        if rng is None:
+            rng = model.default_rng
 
-        total_size = kwargs.pop("total_size", None)
+        _, dims, _ = _validate_shape_dims_size(dims=dims)
```

Comment:
Can we move all this functionality into a separate function?

Reply:
There would be a lot of kwargs to pass. Maybe it could help with testing, but I don't have much hope that it would improve readability a lot.

Reply:
After all that's just a refactoring.
First I'd like to improve at the level of how the `shape`/`size` variables are casted, sliced and their length measured. We can avoid unnecessary Subtensor/MakeVector operations this way.

Reply:
I was able to simplify the branching logic quite a bit.
With how it is now, I don't think we gain anything by separating it out further, because only two lines above and two lines below would remain.